### PR TITLE
sensord: use interrupts to improve LSM6DS3 timing accuracy

### PR DESCRIPTION
**Pull request by pd0wm**
Continuation of #24469

TODO:

- [x] Switch to gpiochip kernel API for more accurate timestamps
- [ ] Check `DRDY_PULSE_CFG_G` register on `LSM6DS3TR` (non `-C` variant)
- [x] Check why mostly falling edges. Is the pulse too short, and edge type measured after the actual interrupt is triggered?
- [x] Look into what triggered kernel panics while testing. 
  Triggered by starting sensord once with  `GPIOEVENT_REQUEST_BOTH_EDGES` followed by `GPIOEVENT_REQUEST_FALLING_EDGE` as argument for `GPIO_GET_LINEEVENT_IOCTL`.

```
1950 [  790.740280] Unable to handle kernel NULL pointer dereference at virtual address 00000004
1951 [  790.740306] pgd = ffffff88fa314000
1952 [  790.740326] [00000004] *pgd=0000000176de8003, *pud=0000000176de8003, *pmd=0000000000000000
1953 [  790.740457] ------------[ cut here ]------------
1954 [  790.740473] Kernel BUG at ffffff88f74ad0f8 [verbose debug info unavailable]
1955 [  790.740492] Internal error: Oops: 96000005 [#1] PREEMPT SMP
1956 [  790.740509] Modules linked in: wlan(CE) snd_soc_sdm845(E) snd_soc_wcd9xxx(E)
1957 [  790.740564] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G         C  E   4.9.103+ #3
1958 [  790.740582] Hardware name: Qualcomm Technologies, Inc. sda845 v2.1 TurboX-SOM_V01 (DT)
1959 [  790.740600] task: ffffff88f949cb80 task.stack: ffffff88f9480000
1960 [  790.740633] PC is at msm_gpio_dirconn_handler+0x30/0xb0
1961 [  790.740658] LR is at generic_handle_irq+0x34/0x50
1962 [  790.740676] pc : [<ffffff88f74ad0f8>] lr : [<ffffff88f71231ac>] pstate: 804001c5
```

### CODE REVIEW

File: common/gpio.cc [hunk 1]
```diff
@@ -30,3 +34,33 @@ int gpio_set(int pin_nr, bool high) {
   }
   return util::write_file(pin_val_path, (void*)(high ? "1" : "0"), 1);
 }
+
+int gpiochip_get_ro_value_fd(int pin_nr, EdgeType etype) {
```
Comment (adeebshihadeh) [inline common/gpio.cc:38]: edge type isn't used

File: common/gpio.cc [hunk 2]
```diff
@@ -30,3 +34,33 @@ int gpio_set(int pin_nr, bool high) {
...
+    LOGE("Error opening gpiochip fd")
+    return -1;
+  }
+
+  // Setup event
+  struct gpioevent_request rq;
+  rq.lineoffset = pin_nr;
+  rq.handleflags = GPIOHANDLE_REQUEST_INPUT;
+
+  // rq.eventflags = (etype == EdgeType::Rising) ? GPIOEVENT_REQUEST_RISING_EDGE : GPIOEVENT_REQUEST_FALLING_EDGE;
```
Comment (adeebshihadeh) [inline common/gpio.cc:50]: let's remove this line, comment sufficiently explains it

File: common/gpio.h [hunk 3]
```diff
@@ -8,14 +8,29 @@
...
+  #define GPIO_BMX_ACCEL_INT    0
+  #define GPIO_BMX_GYRO_INT     0
+  #define GPIO_BMX_MAGN_INT     0
+  #define GPIO_LSM_INT          0
 #endif
 
+enum EdgeType {
+  Rising,
+  Falling,
+};
```
Comment (adeebshihadeh) [inline common/gpio.h:31]: unused

File: selfdrive/sensord/tests/test_sensord.py [hunk 4]
```diff
@@ -55,24 +97,137 @@ def setUpClass(cls):
...
-        # Filter out unset events
+        # filter unset events (bmx magn)
         if measurement.version == 0:
           continue
         seen.add((str(measurement.source), measurement.which()))
 
     self.assertIn(seen, SENSOR_CONFIGURATIONS)
 
+  @with_processes(['sensord'])
+  def test_lsm6ds3_100Hz(self):
```
Comment (adeebshihadeh) [inline selfdrive/sensord/tests/test_sensord.py:114]: let's also add a test case to ensure interrupts aren't coming in after sensord dies

File: selfdrive/sensord/tests/test_sensord.py [hunk 5]
```diff
@@ -2,12 +2,15 @@
...
+from collections import namedtuple
 
 import cereal.messaging as messaging
+from cereal import log
 from system.hardware import TICI
 from selfdrive.test.helpers import with_processes
-
-TEST_TIMESPAN = 10
+from selfdrive.manager.process_config import managed_processes
+from smbus2 import SMBus
```
Comment (adeebshihadeh) [inline selfdrive/sensord/tests/test_sensord.py:13]: add this to the top group of imports, import convention is stdlib + third party then openpilot imports

**Comment by gast04** [comment]
histogram plot of before and after interrupt based lsm6ds3

before:
basic_plotC

after:
ints_plotC

x axis is the time difference between two acceleration measurements of the lsm6ds3 sensor in ns (*1e7 -> 1 on the scale is equal to 10ms which is one period)

File: selfdrive/sensord/sensors_qcom2.cc [hunk 6]
```diff
@@ -24,6 +26,87 @@
...
+      auto orphan = orphanage.newOrphan<cereal::SensorEventData>();
+      auto event = orphan.get();
+      if (!sensor->get_event(event)) {
+        continue;
+      }
+
+      if (ts - init_ts < measure_delay) {
+        // filter first values (0.5sec) as those may contain inaccuracies
+        continue;
+      }
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors_qcom2.cc:84]: this should go in its own PR

File: selfdrive/sensord/sensors_qcom2.cc [hunk 7]
```diff
@@ -102,11 +197,22 @@ int sensor_loop() {
       sensors[i]->get_event(event);
     }
 
-    pm.send("sensorEvents", msg);
+    if (nanos_since_boot() - init_ts < measure_delay) {
+      // filter first values (0.5sec) as those may contain inaccuracies
+      continue;
+    }
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors_qcom2.cc:203]: and this too

File: selfdrive/sensord/sensors_qcom2.cc (hunk 6 above)
Comment (gast04) [reply selfdrive/sensord/sensors_qcom2.cc]: removed

File: selfdrive/sensord/sensors_qcom2.cc (hunk 7 above)
Comment (gast04) [reply selfdrive/sensord/sensors_qcom2.cc]: removed

File: common/gpio.cc [hunk 8]
```diff
@@ -30,3 +30,36 @@ int gpio_set(int pin_nr, bool high) {
   }
   return util::write_file(pin_val_path, (void*)(high ? "1" : "0"), 1);
 }
+
+int gpiochip_get_ro_value_fd(int pin_nr) {
+
+  // Assumed that all interrupt pins are unexported and rights are given to
+  // read from gpiochip0.
+  std::string gpiochip_path = "/dev/gpiochip" + std::to_string(GPIOCHIP_INT);
```
Comment (adeebshihadeh) [inline common/gpio.cc:38]: this should be passed in

File: common/gpio.cc [hunk 9]
```diff
@@ -30,3 +30,36 @@ int gpio_set(int pin_nr, bool high) {
...
+  struct gpioevent_request rq;
+  rq.lineoffset = pin_nr;
+  rq.handleflags = GPIOHANDLE_REQUEST_INPUT;
+
+  /* Requesting both edges as the data ready pulse from the lsm6ds sensor is
+     very short(75us) and is mostly detected as falling edge instead of rising.
+     So if it is detected as rising the following falling edge is skipped. */
+  rq.eventflags = GPIOEVENT_REQUEST_BOTH_EDGES;
+
+  strncpy(rq.consumer_label, "sensord", std::size(rq.consumer_label) - 1);
```
Comment (adeebshihadeh) [inline common/gpio.cc:55]: this should be also be passed in

File: selfdrive/sensord/tests/test_sensord.py [hunk 10]
```diff
@@ -55,24 +115,162 @@ def setUpClass(cls):
...
+    # check if the interrupts got disabled
+    with SMBus(SENSOR_BUS, force=True) as bus:
+      int1_ctrl_reg = bus.read_byte_data(I2C_ADDR_LSM, 0x0D)
+      assert int1_ctrl_reg == 0, "Interrupts not disabled!"
+
+    # read /proc/interrupts to verify no more interrupts are received
+    state_one = get_proc_interrupts(LSM_INT_GPIO)
+    time.sleep(1)
+    state_two = get_proc_interrupts(LSM_INT_GPIO)
+    assert state_one == state_two, "Interrupts received after sensord stop!"
```
Comment (adeebshihadeh) [inline selfdrive/sensord/tests/test_sensord.py:271]: do this check above when you verify that the interrupts are enabled

File: selfdrive/sensord/tests/test_sensord.py [hunk 11]
```diff
@@ -55,24 +115,162 @@ def setUpClass(cls):
...
+    for event in events:
+      for measurement in event.sensorEvents:
+
+        # filter unset events (bmx magn)
+        if measurement.version == 0:
+          continue
+
+        # negative values might occur, as non interrupt packages created
+        tdiffs.append(abs(event.logMonoTime - measurement.timestamp))
+        # before the sensor is read
```
Comment (adeebshihadeh) [inline selfdrive/sensord/tests/test_sensord.py:199]: let's explicitly check the interrupts ones here and make sure that the logMonoTime is after the sample ts. it's ok to just hardcode the list

Comment (adeebshihadeh) [reply selfdrive/sensord/tests/test_sensord.py:199]: oh I see that's a different test case
